### PR TITLE
Stop atlas worker when not enough funds to pay for gas

### PR DESCRIPTION
**Pull request by vanruch**
# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective and that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have tested my changes locally or in the DigitalOcean droplet 
- [x] I have made sure these changes really fix/add new feature as was intended
- [x] I took care of possible corner cases

### CODE REVIEW

File: src/utils/web3_tools.js [hunk 1]
```diff
@@ -116,3 +117,11 @@ export function link(contract, name, library) {
   contract.bytecode = contract.bytecode.replace(pattern, address);
 }
 
+export async function getBalance(web3, address) {
+  return new BN(await web3.eth.getBalance(address));
```
Comment (spherefoundry) [inline src/utils/web3_tools.js]: Functions should return POJSO as a rule (In this case a string). For context: we already had a lot of trouble when some functions returned a BN and some others have returned a string.

File: src/utils/web3_tools.js [hunk 2]
```diff
@@ -116,3 +117,11 @@ export function link(contract, name, library) {
...
+}
+
+export function maximalGasPrice() {
+  const GAS_LIMIT = '8000000';
+  return new BN(GAS_LIMIT).mul(new Web3().utils.toWei(new BN(config.defaultGasPrice), 'gwei'));
```
Comment (spherefoundry) [inline src/utils/web3_tools.js]: Same here. For calculations it's ok to wrap into BN. But then the result should be converted back into string.

Comment (spherefoundry) [reply src/utils/web3_tools.js]: ⚠️  Magic number warning ⚠️ Why not use the DEFAULT_GAS? All transactions are performed with it by default.

File: src/workers/atlas_worker.js [hunk 3]
```diff
@@ -36,6 +37,10 @@ export default class AtlasWorker extends PeriodicWorker {
   }
 
   async periodicWork() {
+    if (!await this.checkIfEnoughFundsToPayForGas()) {
+      await this.addLog('Not enough funds to pay for gas');
```
Comment (spherefoundry) [inline src/workers/atlas_worker.js]: Oh, this is problematic. Basically we will spam MongoDB with this message every second. Totalling at 86400 entries a day until this is resolved. I think the check should be performed on every work loop. But the log should only happen on change enough > not-enough and not-enough > enough.

File: src/workers/atlas_worker.js [hunk 4]
```diff
@@ -57,6 +62,10 @@ export default class AtlasWorker extends PeriodicWorker {
     await this.dataModelEngine.cleanupBundles();
   }
 
+  async checkIfEnoughFundsToPayForGas() {
+    return (await getBalance(this.web3)).gt(maximalGasPrice());
```
Comment (spherefoundry) [inline src/workers/atlas_worker.js]: Move to web3 tools?

File: test/utils/web3_tools.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  it('maximalGasPrice returns correct value', () => {
+    expect(maximalGasPrice()).to.equal('23500000000000000');
+  });
+
+  describe('checkIfEnoughFundsToPayForGas', () => {
+    let getBalanceMock;
+
+    before(() => {
+      getBalanceMock = sinon.stub(web3.eth, 'getBalance');
```
Comment (vanruch) [inline test/utils/web3_tools.js]: Will be replaced with mock web3 after merge with #272

File: src/utils/web3_tools.js [hunk 6]
```diff
@@ -136,3 +137,17 @@ export async function waitForChainSync(web3, timeoutInSeconds, iterationCallback
     await sleep(timeoutInSeconds);
   }
 }
+
+export async function getBalance(web3, address) {
+  return await web3.eth.getBalance(address);
+}
+
+export function maximalGasPrice() {
+  return new BN(DEFAULT_GAS).mul(new Web3().utils.toWei(new BN(config.defaultGasPrice), 'gwei'))
```
Comment (Anze1m) [inline src/utils/web3_tools.js]: `new Web3()`? 😮

File: src/utils/web3_tools.js [hunk 7]
```diff
@@ -9,6 +9,7 @@ This Source Code Form is “Incompatible With Secondary Licenses”, as defined
 
 import Web3 from 'web3';
 import config from '../../config/config';
+import BN from 'bn.js';
 
 export const DEFAULT_GAS = 4700000;
```
Comment (Anze1m) [inline src/utils/web3_tools.js]: Why isn't it in config?

**Comment by vanruch** [comment]
@spherefoundry CR please 
image

**Comment by spherefoundry** [comment]
:+1: for the improved tests

**Pull request merged**